# HG changeset patch
# User Rik <octave@nomad.inbox5.com>
# Date 1315179976 25200
#      Sun Sep 04 16:46:16 2011 -0700
# Node ID 52e4aa30d5b249e8b76fdbbc1a9e7c0e15b28359
# Parent  5553412c661410e05fa3cb4f7f660dbcf891a8c2
logm.m: Return real matrix when all eigenvalues are real (Bug #32121).

* logm.m: Remove complex numbers of order eps() which may have entered
return value through numeric roundoff.

diff --git a/scripts/linear-algebra/logm.m b/scripts/linear-algebra/logm.m
--- a/scripts/linear-algebra/logm.m
+++ b/scripts/linear-algebra/logm.m
@@ -58,21 +58,24 @@ function [s, iters] = logm (A, opt_iters
   endif
 
   [u, s] = schur (A);
 
   if (isreal (A))
     [u, s] = rsf2csf (u, s);
   endif
 
-  if (any (diag (s) < 0))
+  eigv = diag (s);
+  if (any (eigv < 0))
     warning ("Octave:logm:non-principal",
              "logm: principal matrix logarithm is not defined for matrices with negative eigenvalues; computing non-principal logarithm");
   endif
 
+  real_eig = all (eigv >= 0);
+
   k = 0;
   ## Algorithm 11.9 in "Function of matrices", by N. Higham
   theta = [0, 0, 1.61e-2, 5.38e-2, 1.13e-1, 1.86e-1, 2.6429608311114350e-1];
   p = 0;
   m = 7;
   while (k < opt_iters)
     tau = norm (s - eye (size (s)),1);
     if (tau <= theta (7))
@@ -95,16 +98,21 @@ function [s, iters] = logm (A, opt_iters
   s = s - eye (size (s));
 
   if (m > 1)
     s = logm_pade_pf (s, m);
   endif
 
   s = 2^k * u * s * u';
 
+  ## Remove small complex values (O(eps)) which may have entered calculation
+  if (real_eig)
+    s = real (s);
+  endif
+
   if (nargout == 2)
     iters = k;
   endif
 
 endfunction
 
 ################## ANCILLARY FUNCTIONS ################################
 ######  Taken from the mfttoolbox (GPL 3) by D. Higham.
